### PR TITLE
Co tracker

### PR DESCRIPTION
**Pull request by velocitatem**
Work in Progress ✔️

### CODE REVIEW

**Comment by velocitatem** [comment]
We could run the notebook on the big screen at the lab? Since it integrates with the computer's camera, could be nice to see a semi-live demo.

**Comment by velocitatem** [comment]
+ [x] Add comments to the code - mapping the inputs from the paper

**Comment by velocitatem** [comment]
https://viper.cs.columbia.edu/ Is the other model I mentioned at the end.

File: cotracker/cotracker-simple-implementation.ipynb
Comment (dsanmart) [inline cotracker/cotracker-simple-implementation.ipynb:1]: Could you run the notebook and commit the outputs? This way everyone can take a quick look at them without having to do the whole setup.

Comment (velocitatem) [reply cotracker/cotracker-simple-implementation.ipynb:1]: Sounds good! Will do 👍🏻

File: cotracker/README.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+1. **Optical Flow**: This method estimates the motion of every single point in a video frame but only for a very short period.
+2. **Point Tracking**: This focuses on tracking specific points over an extended period but does it independently, without considering the relationship between points.
+
+Both methods have their limitations. For example, they struggle with tracking points that get temporarily hidden (occluded) or are part of the same object.
+
+## High Level Overview
+
+CoTracker is built on the idea that points in a video are often correlated. For example, if you're tracking the wheels of a car, it's likely that if one wheel moves, the other will too. The authors argue that by ignoring these correlations, we're missing out on improving the accuracy of tracking. So, they propose a new architecture that tracks multiple points jointly throughout an entire video sequence. This is a game-changer because it allows the model be more informed of the correlations between points that are being tracked.
+
+```mermaid
```
Comment (dsanmart) [inline cotracker/README.md:30]: If you are going into how the architecture of Cotracker works. Please also include figures 2 and 3 from the paper for better and more accurate understanding.

**Review by dsanmart — CHANGES_REQUESTED** [review]
Great content! Thanks for the contribution @velocitatem and all the explanations in the README. I added some minor feedback

File: Makefile [hunk 2]
```diff
@@ -1,10 +1,10 @@
 TOPIC ?= gans
+TOPIC ?= cotracker
```
Comment (vtwoptwo) [inline Makefile]: here there is no need to add another default. I just added it in case someone runs the make command without setting a different TOPIC variable

File: cotracker/README.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+- **Output: Visibility Flags**: $\hat{v}_{ti}$
+  The estimated visibility flags indicate whether each point is visible or occluded in each frame.
+
+
+The technical backbone of CoTracker is a [transformer network](https://mchromiak.github.io/articles/2017/Sep/12/Transformer-Attention-is-all-you-need/), a type of neural network that's particularly good at handling sequences and relationships within them. The transformer iteratively refines its estimates of where points are in each frame, effectively learning from the video as it goes along. What's even cooler is that this isn't just for short clips; CoTracker is designed to handle long videos by using a sliding-window approach. This means it can update its tracking in real-time as it processes the video, making it incredibly flexible and scalable.
+
+## Combination with Viper
+
+CoTracker could serve as one of the modules in ViperGPT's API. For example, a query like "Track the movement of the red car and the blue car and tell me when they are closest" could be answered by combining CoTracker's tracking capabilities with ViperGPT's code generation and reasoning.
+
```
Comment (vtwoptwo) [inline cotracker/README.md:75]:
```suggestion
Example prompts:
```

File: cotracker/README.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
+---
+Paper: https://arxiv.org/pdf/2307.07635.pdf
+Official Website: https://co-tracker.github.io/
+Demo: https://huggingface.co/spaces/facebook/cotracker
+---
+# CoTracker
+
+A new state of the art architecture for optical flow and particle tracking proposed by meta. The key innovation in this project is the use of Transformer Architecture and the idea that tracking multiple particles concurrently, can help boos overall accuracy.
```
Comment (vtwoptwo) [inline cotracker/README.md]:
```suggestion
A new state-of-the-art architecture for optical flow and particle tracking proposed by Meta. The key innovation in this project is the use of Transformer Architecture and the idea that tracking multiple particles concurrently, can help boos overall accuracy.
```

Comment (vtwoptwo) [inline cotracker/README.md]: The key innovation in this project is the use of Transformer Architecture and the idea that tracking multiple particles concurrently, can help boos overall accuracy. -> This sentence is a bit weird maybe rephrase it...?

File: cotracker/cotracker-simple-implementation.ipynb [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+   "id": "58a3f066",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "output_filename=\"my_video_tracked\"\n",
+    "file = vis.visualize(\n",
+    "    video_tensor,\n",
+    "    tracks=tracks,\n",
+    "    visibility=visibility,\n",
+    "    filename=output_filename, # name of file t\n",
```
Comment (vtwoptwo) [inline cotracker/cotracker-simple-implementation.ipynb]:
```suggestion
    "    filename=output_filename, # name of output file\n",
```

File: cotracker/cotracker-simple-implementation.ipynb
Comment (vtwoptwo) [inline cotracker/cotracker-simple-implementation.ipynb:1]: In general cool! Maybe comment each cell what each step is.

File: cotracker/requirements.txt
Comment (vtwoptwo) [inline cotracker/requirements.txt:1]: versions for each dependency

**Review by vtwoptwo — CHANGES_REQUESTED** [review]
Awesome! Looks great. Ping me when u make the small changes and other than that LGTM!

**Comment by dsanmart** [comment]
It seems like the png files for the figures have not been committed

**Review by vtwoptwo — COMMENTED** [review]
lgtm

**Review by vtwoptwo — APPROVED** [review]
